File: backend/app/routers/workspaces.py
"""Workspace CRUD and membership.

A workspace is the tenant that owns custom domains, branding and members. It is
entirely additive: a user with no workspace, and a document with no
`workspace_id`, behave exactly as they did before this existed.
"""

from fastapi import APIRouter, Depends, HTTPException, Request
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.database import get_database
from app.limiter import limiter
from app.models.user import User
from app.models.workspace import Workspace
from app.routers.auth import require_user
from app.schemas.workspace import (
    BrandingPayload,
    MemberAddRequest,
    MemberListResponse,
    MemberResponse,
    MemberRoleRequest,
    SettingsPayload,
    WorkspaceCreate,
    WorkspaceListResponse,
    WorkspaceResponse,
    WorkspaceUpdate,
)
from app.services import workspace as ws_service

router = APIRouter(prefix="/api/v1/workspaces", tags=["workspaces"])


def get_db() -> AsyncIOMotorDatabase:
    return get_database()


def _to_response(workspace: Workspace, role: str) -> WorkspaceResponse:
    b = workspace.branding
    s = workspace.settings
    return WorkspaceResponse(
        id=workspace.id,
        name=workspace.name,
        owner_id=workspace.owner_id,
        created_at=workspace.created_at,
        updated_at=workspace.updated_at,
        branding=BrandingPayload(
            site_name=b.site_name,
            favicon_url=b.favicon_url,
            logo_url=b.logo_url,
            accent_color=b.accent_color,
            hide_markdrop_branding=b.hide_markdrop_branding,
        ),
        settings=SettingsPayload(
            viewer_chrome=s.viewer_chrome,
            require_auth_to_view=s.require_auth_to_view,
        ),
        role=role,  # type: ignore[arg-type]
    )


@router.get("", response_model=WorkspaceListResponse)
@limiter.limit("60/minute")
async def list_workspaces(
    request: Request,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    pairs = await ws_service.list_for_user(db, user.id)
    return WorkspaceListResponse(
        workspaces=[_to_response(w, role) for w, role in pairs]
    )


@router.post("", response_model=WorkspaceResponse, status_code=201)
@limiter.limit("10/minute")
async def create_workspace(
    request: Request,
    data: WorkspaceCreate,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    workspace = await ws_service.create_workspace(db, data.name, user)
    return _to_response(workspace, "owner")


@router.get("/{workspace_id}", response_model=WorkspaceResponse)
@limiter.limit("120/minute")
async def get_workspace(
    request: Request,
    workspace_id: str,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    role = await ws_service.require_role(db, workspace_id, user.id, "viewer")
    workspace = await ws_service.get_workspace(db, workspace_id)
    if workspace is None:
        raise HTTPException(status_code=404, detail="Workspace not found")
    return _to_response(workspace, role)


@router.put("/{workspace_id}", response_model=WorkspaceResponse)
@limiter.limit("60/minute")
async def update_workspace(
    request: Request,
    workspace_id: str,
    data: WorkspaceUpdate,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    # Branding and viewer settings change what every visitor sees, so they sit
    # behind admin rather than member.
    role = await ws_service.require_role(db, workspace_id, user.id, "admin")
    workspace = await ws_service.update_workspace(
        db,
        workspace_id,
        name=data.name,
        branding=data.branding.model_dump() if data.branding is not None else None,
        settings=data.settings.model_dump() if data.settings is not None else None,
    )
    return _to_response(workspace, role)


# ── Members ───────────────────────────────────────────────────────────────────


@router.get("/{workspace_id}/members", response_model=MemberListResponse)
@limiter.limit("120/minute")
async def list_members(
    request: Request,
    workspace_id: str,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    await ws_service.require_role(db, workspace_id, user.id, "viewer")
    members = await ws_service.list_members(db, workspace_id)
    return MemberListResponse(
        members=[
            MemberResponse(
                user_id=m.user_id, role=m.role, email=m.email, name=m.name,
                created_at=m.created_at,
            )
            for m in members
        ]
    )


@router.post("/{workspace_id}/members", response_model=MemberResponse, status_code=201)
@limiter.limit("30/minute")
async def add_member(
    request: Request,
    workspace_id: str,
    data: MemberAddRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    await ws_service.require_role(db, workspace_id, user.id, "admin")
    member = await ws_service.add_member(db, workspace_id, str(data.email), data.role)
    return MemberResponse(
        user_id=member.user_id, role=member.role, email=member.email,
        name=member.name, created_at=member.created_at,
    )


@router.put("/{workspace_id}/members/{member_user_id}", status_code=204)
@limiter.limit("30/minute")
async def set_member_role(
    request: Request,
    workspace_id: str,
    member_user_id: str,
    data: MemberRoleRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    await ws_service.require_role(db, workspace_id, user.id, "admin")
    await ws_service.set_member_role(db, workspace_id, member_user_id, data.role)


@router.delete("/{workspace_id}/members/{member_user_id}", status_code=204)
@limiter.limit("30/minute")
async def remove_member(
    request: Request,
    workspace_id: str,
    member_user_id: str,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: User = Depends(require_user),
):
    # Leaving is always allowed; removing someone else needs admin.
    if member_user_id != user.id:
        await ws_service.require_role(db, workspace_id, user.id, "admin")
    else:
        await ws_service.require_role(db, workspace_id, user.id, "viewer")
    await ws_service.remove_member(db, workspace_id, member_user_id)

File: frontend/src/lib/workspaces.ts
/**
 * Workspace, domain, member and folder API.
 *
 * Kept out of lib/api.ts: that module is loaded by every public document page,
 * and none of this is reachable without a session.
 */
import { API_BASE, type MyDocListItem, type MyDocListResponse } from "@/lib/api";

const TOKEN_KEY = "markdrop_token";

function authHeaders(): Record<string, string> {
  const t = typeof window === "undefined" ? null : localStorage.getItem(TOKEN_KEY);
  return t ? { Authorization: `Bearer ${t}` } : {};
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    ...init,
    cache: "no-store",
    headers: { "Content-Type": "application/json", ...authHeaders(), ...(init?.headers || {}) },
  });
  if (!res.ok) {
    const body = await res.json().catch(() => null);
    // FastAPI validation errors arrive as a list of objects; surfacing "[object
    // Object]" to a user is worse than saying nothing useful at all.
    const detail = body?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : Array.isArray(detail)
          ? detail[0]?.msg || "That value isn't valid."
          : `Request failed (${res.status})`;
    throw new Error(message);
  }
  return res.status === 204 ? (undefined as T) : res.json();
}

// ── Types ─────────────────────────────────────────────────────────────────────

export type Role = "owner" | "admin" | "member" | "viewer";
export type DomainKind = "app" | "cdn";
export type ViewerChrome = "full" | "minimal" | "none";

export interface Branding {
  site_name: string | null;
  favicon_url: string | null;
  logo_url: string | null;
  accent_color: string | null;
  hide_markdrop_branding: boolean;
}

export interface WorkspaceSettings {
  viewer_chrome: ViewerChrome;
  require_auth_to_view: boolean;
}

export interface Workspace {
  id: string;
  name: string;
  owner_id: string;
  created_at: string;
  updated_at: string;
  branding: Branding;
  settings: WorkspaceSettings;
  role: Role;
  /** So the settings header is correct without fetching three collections. */
  member_count: number;
  pending_invite_count: number;
  domain_count: number;
}

export interface Member {
  user_id: string;
  role: Role;
  email: string | null;
  name: string | null;
  created_at: string;
}

export interface Domain {
  id: string;
  host: string;
  kind: DomainKind;
  status: "pending" | "verified" | "failed";
  created_at: string;
  verified_at: string | null;
  last_checked_at: string | null;
  last_error: string | null;
  attached: boolean;
  dns_record_name: string;
  dns_record_type: string;
  dns_record_value: string;
  dns_target_name: string;
  dns_target_type: string;
  dns_target_value: string;
  warning: string | null;
}

export interface Folder {
  id: string;
  workspace_id: string;
  name: string;
  parent_id: string | null;
  created_at: string;
  updated_at: string;
}

// ── Workspaces ────────────────────────────────────────────────────────────────

export const listWorkspaces = () =>
  request<{ workspaces: Workspace[] }>("/api/v1/workspaces").then((r) => r.workspaces);

export const createWorkspace = (name: string) =>
  request<Workspace>("/api/v1/workspaces", { method: "POST", body: JSON.stringify({ name }) });

export const getWorkspace = (id: string) => request<Workspace>(`/api/v1/workspaces/${id}`);

/** Delete a workspace. Owner only, and the exact name must be typed back.
 *  Shared documents are released to their owners, never deleted. */
export const deleteWorkspace = (id: string, confirmName: string) =>
  request<{ documents_released: number; domains_removed: number; members_removed: number }>(
    `/api/v1/workspaces/${id}`,
    { method: "DELETE", body: JSON.stringify({ confirm_name: confirmName }) }
  );

export const updateWorkspace = (
  id: string,
  data: { name?: string; branding?: Branding; settings?: WorkspaceSettings }
) => request<Workspace>(`/api/v1/workspaces/${id}`, { method: "PUT", body: JSON.stringify(data) });

// ── Members ───────────────────────────────────────────────────────────────────

export const listMembers = (id: string) =>
  request<{ members: Member[] }>(`/api/v1/workspaces/${id}/members`).then((r) => r.members);

export const setMemberRole = (id: string, userId: string, role: Exclude<Role, "owner">) =>
  request<void>(`/api/v1/workspaces/${id}/members/${userId}`, {
    method: "PUT",
    body: JSON.stringify({ role }),
  });

export const removeMember = (id: string, userId: string) =>
  request<void>(`/api/v1/workspaces/${id}/members/${userId}`, { method: "DELETE" });

// ── Invitations ───────────────────────────────────────────────────────────────
//
// There is no "add member" call. Joining a workspace is the invitee's decision,
// so it always goes through an invitation they have to accept.

export type InviteStatus = "pending" | "accepted" | "declined" | "revoked" | "expired";

export interface Invitation {
  id: string;
  email: string;
  role: Role;
  status: InviteStatus;
  created_at: string;
  expires_at: string;
  invited_by_name: string | null;
  responded_at: string | null;
}

export interface InvitePreview {
  workspace_name: string;
  role: Role;
  email: string;
  invited_by_name: string | null;
  status: InviteStatus;
  expires_at: string;
  signed_in_as: string | null;
  email_matches: boolean;
  already_member: boolean;
}

export const listInvitations = (id: string) =>
  request<{ invitations: Invitation[] }>(`/api/v1/workspaces/${id}/invitations`).then(
    (r) => r.invitations
  );

export const inviteMember = (id: string, email: string, role: Exclude<Role, "owner">) =>
  request<Invitation>(`/api/v1/workspaces/${id}/invitations`, {
    method: "POST",
    body: JSON.stringify({ email, role }),
  });

export const revokeInvitation = (id: string, inviteId: string) =>
  request<void>(`/api/v1/workspaces/${id}/invitations/${inviteId}`, { method: "DELETE" });

export const previewInvite = (token: string) =>
  request<InvitePreview>(`/api/v1/invites/${encodeURIComponent(token)}`);

export const acceptInvite = (token: string) =>
  request<{ workspace_id: string; workspace_name: string; role: Role }>(
    `/api/v1/invites/${encodeURIComponent(token)}/accept`,
    { method: "POST" }
  );

export const declineInvite = (token: string) =>
  request<void>(`/api/v1/invites/${encodeURIComponent(token)}/decline`, { method: "POST" });

// ── Branding assets ───────────────────────────────────────────────────────────

/** Upload a favicon or logo. The server re-encodes it to PNG and returns the
 *  hosted URL; it is not saved onto the workspace until the form is submitted. */
export async function uploadBrandingAsset(
  id: string,
  kind: "favicon" | "logo",
  file: File
): Promise<string> {
  const body = new FormData();
  body.append("file", file);
  const res = await fetch(`${API_BASE}/api/v1/workspaces/${id}/branding/${kind}`, {
    method: "POST",
    // No Content-Type: the browser has to set the multipart boundary itself.
    headers: authHeaders(),
    body,
  });
  if (!res.ok) {
    const b = await res.json().catch(() => null);
    throw new Error(typeof b?.detail === "string" ? b.detail : "Upload failed");
  }
  return (await res.json()).url as string;
}

// ── Domains ───────────────────────────────────────────────────────────────────

export const listDomains = (id: string) =>
  request<{ domains: Domain[] }>(`/api/v1/workspaces/${id}/domains`).then((r) => r.domains);

export interface DnsProviderHint {
  detected: boolean;
  provider_id: string | null;
  provider_name: string | null;
  panel_url: string | null;
  host_field: string | null;
  record_host: string | null;
  target_host: string | null;
  note: string | null;
  nameservers: string[];
}

/** Who runs this domain's DNS, so the setup steps can use their wording.
 *  Costs a live DNS lookup, so it is fetched only while the steps are shown. */
export const dnsProviderHint = (id: string, domainId: string) =>
  request<DnsProviderHint>(`/api/v1/workspaces/${id}/domains/${domainId}/provider`);

export const addDomain = (id: string, host: string, kind: DomainKind) =>
  request<Domain>(`/api/v1/workspaces/${id}/domains`, {
    method: "POST",
    body: JSON.stringify({ host, kind }),
  });

export const verifyDomain = (id: string, domainId: string) =>
  request<Domain>(`/api/v1/workspaces/${id}/domains/${domainId}/verify`, { method: "POST" });

export const attachDomain = (id: string, domainId: string) =>
  request<Domain>(`/api/v1/workspaces/${id}/domains/${domainId}/attach`, { method: "POST" });

export const removeDomain = (id: string, domainId: string) =>
  request<void>(`/api/v1/workspaces/${id}/domains/${domainId}`, { method: "DELETE" });

// ── Folders ───────────────────────────────────────────────────────────────────

export const listFolders = (id: string) =>
  request<{ folders: Folder[] }>(`/api/v1/workspaces/${id}/folders`).then((r) => r.folders);

export const createFolder = (id: string, name: string, parentId?: string | null) =>
  request<Folder>(`/api/v1/workspaces/${id}/folders`, {
    method: "POST",
    body: JSON.stringify({ name, parent_id: parentId ?? null }),
  });

export const renameFolder = (id: string, folderId: string, name: string) =>
  request<Folder>(`/api/v1/workspaces/${id}/folders/${folderId}`, {
    method: "PUT",
    body: JSON.stringify({ name }),
  });

export const deleteFolder = (id: string, folderId: string) =>
  request<{ unfiled_documents: number }>(`/api/v1/workspaces/${id}/folders/${folderId}`, {
    method: "DELETE",
  });

// ── Shared library ────────────────────────────────────────────────────────────
//
// A document is private until its owner shares it here. Nothing in this API can
// surface a document that was never shared: the listing matches on workspace_id,
// which a private document does not have.

export const listLibrary = (
  id: string,
  opts: { page?: number; limit?: number; q?: string; kind?: string; folderId?: string | null; unfiled?: boolean } = {}
) => {
  const p = new URLSearchParams();
  p.set("page", String(opts.page ?? 1));
  p.set("limit", String(opts.limit ?? 20));
  if (opts.q) p.set("q", opts.q);
  if (opts.kind) p.set("kind", opts.kind);
  if (opts.folderId) p.set("folder_id", opts.folderId);
  if (opts.unfiled) p.set("unfiled", "true");
  return request<MyDocListResponse>(`/api/v1/workspaces/${id}/documents?${p}`);
};

export const libraryCounts = (id: string) =>
  request<Record<string, number>>(`/api/v1/workspaces/${id}/documents/counts`);

/** Share a document you own. Everyone in the workspace can then read it, and
 *  members and admins can edit it — the UI must say so before calling this. */
export const shareToWorkspace = (id: string, documentId: string, folderId?: string | null) =>
  request<MyDocListItem>(`/api/v1/workspaces/${id}/documents`, {
    method: "POST",
    body: JSON.stringify({ document_id: documentId, folder_id: folderId ?? null }),
  });

/** Unshare. The document itself is untouched and returns to its owner. */
export const unshareFromWorkspace = (id: string, documentId: string) =>
  request<void>(`/api/v1/workspaces/${id}/documents/${documentId}`, { method: "DELETE" });

export const fileDocument = (id: string, documentId: string, folderId: string | null) =>
  request<void>(`/api/v1/workspaces/${id}/documents/${documentId}/folder`, {
    method: "PUT",
    body: JSON.stringify({ folder_id: folderId }),
  });

/** Roles that can act, by capability. Mirrors ROLE_RANK on the server — the UI
 *  hides what the API would refuse, rather than inventing its own rules. */
export const ROLE_RANK: Record<Role, number> = { viewer: 0, member: 1, admin: 2, owner: 3 };
export const can = (role: Role | undefined, required: Role) =>
  role !== undefined && ROLE_RANK[role] >= ROLE_RANK[required];
